### PR TITLE
Add a way to check for white space specifications and alert the user to their presence.

### PR DESCRIPTION
**Pull request by PompeiaPaetenari**
We found that users could alter the file and project settings to enforce different spacings and tabs on others within the project space and needed a way to ensure that it causes an error during the linting phase to prevent it.  These settings can be provided on a per-user basis within XCode and should not be done on the project or file level, as they often get missed as part of a PR and then influence others working on the same code.

### CODE REVIEW

File: Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * the License.
+ */
+
+import Foundation
+
+public func checkForWhiteSpaceSpecifications(_ project: Project, errorReporter: ErrorReporter) -> Int32 {
+  var errors: [String] = []
+  var scriptResult = EX_OK
+
+  for namedGroup in project.groups.keys {
```
Comment (Alex-Ozun) [inline Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift]: Should we generate separate error for each whitespace specification? 
I think it will be very helpful when you have multiple specifications in a single group or file. Otherwise, removing one specification will result in the same error as before removal and may give false impression, that nothing has changed.

You should also be able to iterate over `values` to avoid `guard let group = project.groups[namedGroup] else { continue }`

File: Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      errors.append("\(errorReporter.reportKind.logEntry) Group item (\(group.id)) contains whitespace specification.\n")
+
+      scriptResult = errorReporter.reportKind.returnType
+    }
+  }
+
+  for namedFileReference in project.fileReferences.keys {
+    guard let fileReference = project.fileReferences[namedFileReference] else { continue }
+    if fileReference.tabWidth != nil || fileReference.indentWidth != nil || fileReference.lineEnding != nil {
+      errors.append("\(errorReporter.reportKind.logEntry) File “\(fileReference.title)” (\(fileReference.id)) contains whitespace specification.\n")
```
Comment (Alex-Ozun) [inline Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift]: If you insert "`\(project.absolutePathToReference(fileReference)):0:`" before `"\(errorReporter.reportKind.logEntry) File “\(fileReference.title)” (\(fileReference.id)) contains whitespace specification.\n"`
Xcode will show this error inside a relevant file.

File: Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift (hunk 1 above)
Comment (PompeiaPaetenari) [reply Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift]: I started to do that, but think it would result in a ton of spam on the line.  Already, any group or file that has at least one will notify you, and the README tells the user what is checked, so they could look for them.  I guess, if that is too much back-and-forth for someone new to the tool, or getting an update, then we could fill the screen of each instance instead... I am not sure on which is preferred.

Comment (Alex-Ozun) [reply Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift]: Yeah, I think it's fine. Let's try out unified error message first and see if it's good enough.

I'd still ask you to iterate over `project.groups.values` to avoid `guard let group = project.groups[namedGroup] else { continue }`.
Same can be done with `project.fileReferences.values`

File: Makefile [hunk 3]
```diff
@@ -35,7 +35,7 @@ release:
 # It doesn’t work. Xcode release notes say:
 # 	Creating static executables using the new option --static-swift-stdlib does not function correctly. (33861492)
 # so we’ll use their workaround:
-	swift build --configuration release  -Xswiftc -static-stdlib
```
Comment (kalkwarf) [inline Makefile]: If you're going to remove the arguments, we need to update the comments.

See what I mean about comments getting out of date? :-)

File: README.md [hunk 4]
```diff
@@ -154,6 +159,15 @@ inside.
     `allNames`. We sort the groups and files, contatenate them, then compare
     that to the list of everything.
 
+- No White Space Specifications
+
+    XCode has a way to, at the user-level, specify customisations for tabs, spaces,
```
Comment (jakeatoms) [inline README.md]: It's `Xcode`

File: README.md [hunk 5]
```diff
@@ -154,6 +159,15 @@ inside.
...
+    indentation amount, and line endings within `Preferences -> Text Editing ->
+    Indentation`.  However, the user might change these preferences at a file or
+    project level, altering the experience for team-members.  As such, we want to
+    keep the file clean of these preferences and rely on `swiftformat` to shape
```
Comment (jakeatoms) [inline README.md]: Do we want to mention `swiftformat` here? Also, I think we should keep the documentation in the 3rd person.

> ...altering the experience for team-members. Formatting can be enforced using other tools or implementations.

File: Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+@testable import xcprojectlint_package
+import XCTest
+
+class WhiteSpaceSpecificationTests: XCTestCase {
```
Comment (jakeatoms) [inline Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift]: This should be `final`

File: Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  func test_whiteSpaceSpecifiersArePresent_returnsError() {
```
Comment (jakeatoms) [inline Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift]: I think a test of the happy path is also appropriate.

File: TestData/Bad.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -96,7 +96,10 @@
 				D2A90D18203219FC00EBA6AA /* ThisGroupIsEmpty */,
 			);
 			path = Bad;
+			indentWidth = 7;
 			sourceTree = "<group>";
+			tabWidth = 7;
+			usesTabs = 1;
```
Comment (jakeatoms) [inline TestData/Bad.xcodeproj/project.pbxproj]: It's going to be hard to know why the test failed, though. Ideally, we'd have these scenarios broken out into more granular tests so we can validate each logic branch (both for file/project settings, as well as the specific whitespace rules)

File: README.md [hunk 9]
```diff
@@ -164,8 +178,7 @@ and to recipients of software distributed by American Express, You reserve all
 right, title, and interest, if any, in and to Your Contributions. Please [fill
 out the Agreement](https://cla-assistant.io/americanexpress/xcprojectlint).
 
-Please feel free to open pull requests and see
-[CONTRIBUTING.md](./CONTRIBUTING.md) for commit formatting details.
+Please feel free to open pull requests.
```
Comment (jakeatoms) [inline README.md]: Why remove the link to Contributing?

Comment (kalkwarf) [reply README.md]: That file doesn't exist.

Comment (jakeatoms) [reply README.md]: TIL: `CONTRIBUTING.md` is gone

File: Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift (hunk 7 above)
Comment (PompeiaPaetenari) [reply Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift]: Okay, that sounds good.  I had only copied what we had in the other tests to try to be conformant, but that sounds like a better approach.

File: README.md (hunk 9 above)
Comment (PompeiaPaetenari) [reply README.md]: Yep!  I noticed that early on when looking for the formatting details.  In the end, I just went with how we currently format our code and used that on the files I touched.

File: Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift (hunk 6 above)
Comment (PompeiaPaetenari) [reply Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift]: Should I make all the test classes final then?  Currently none of them are.  Or should I leave that as a separate PR?

File: README.md (hunk 5 above)
Comment (PompeiaPaetenari) [reply README.md]: Ah, okay.  I am used to the "software paper 'we' second-person" styling from the tech papers at uni.  Can change it.

File: TestData/Bad.xcodeproj/project.pbxproj (hunk 8 above)
Comment (PompeiaPaetenari) [reply TestData/Bad.xcodeproj/project.pbxproj]: I kept going back and forth on spreading them out so we get a hit on each separate, because of my use of the or logic.  But not 100% sure on how much that needs to be tested out.  I am okay spreading one to each line to show getting a ping on each item.

Comment (kalkwarf) [reply TestData/Bad.xcodeproj/project.pbxproj]: Let's look at it from the perspective of a person adding this tool to their build process.

When you get the error, what does it say? 

When you address the reported error, does the next build succeed?

It seems to me that knowing you had `tabWidth` _and_ `usesTabs` set would be more informative than knowing you had "white space specifiers" set.

The goal would be to make it more likely you'd fix it the first time.

File: Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift (hunk 6 above)
Comment (jakeatoms) [reply Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift]: I suppose we can do that separately

File: TestData/Bad.xcodeproj/project.pbxproj (hunk 8 above)
Comment (PompeiaPaetenari) [reply TestData/Bad.xcodeproj/project.pbxproj]: Okay.  Alex had said the same thing, and I worried it would be too spammy.  But if that is preferred, I will change to display every one of them found.

File: Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift (hunk 6 above)
Comment (PompeiaPaetenari) [reply Tests/xcprojectlint-packageTests/WhiteSpaceSpecificationTests.swift]: https://github.com/americanexpress/xcprojectlint/pull/14

File: Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import Foundation
+
+public func checkForWhiteSpaceSpecifications(_ project: Project, errorReporter: ErrorReporter) -> Int32 {
+  let mapToGroupError: (String, String) -> (String) -> String = { groupID, type in
```
Comment (Alex-Ozun) [inline Sources/xcprojectlint-package/CheckForWhitespaceSpecifications.swift]: nitpick: `toGroupError` and `toFileError` reads a bit better at invocation `group.tabWidth.map(toGroupError(group.id, "tabWidth"))`.
